src/arch/riscv/pagetable_walker.cc

```diff
@@ -242,91 +284,355 @@ Walker::startWalkWrapper()
         // check the next translation request, if it exists
         if (currStates.size()) {
             currState = currStates.front();
-            vaddr = Addr(sext<VADDR_BITS>(currState->req->getVaddr()));
+            vaddr = Addr(sext<SV39_VADDR_BITS>(currState->req->getVaddr()));
             Addr vpn = getVPNFromVAddr(vaddr, currState->satp.mode);
             e = tlb->lookup(vpn, currState->satp.asid, currState->mode,
                             true);
             if (e) {
-               fault = tlb->checkPermissions(currState->status,
-                                             currState->pmode, vaddr,
-                                             currState->mode, e->pte);
+                fault = tlb->checkPermissions(currState->tc,
+                    currState->memaccess, e->vaddr, currState->mode, e->pte);
             }
         } else {
             currState = NULL;
         }
     }
     if (currState && !currState->wasStarted()) {
         if (!e || fault != NoFault)
-            currState->startWalk();
+            currState->walk();
         else
             schedule(startWalkWrapperEvent, clockEdge(Cycles(1)));
     }
 }
 
 Fault
-Walker::WalkerState::startWalk()
+Walker::WalkerState::walkGStage(Addr guest_paddr, Addr& host_paddr)
+{
+    Fault fault = NoFault;
+    curstage = XlateStage::GSTAGE;
+
+
+    // reset gresult in case we were called again
+    // in a two stage translation
+    gresult.reset();
+    gresult.vaddr = guest_paddr;
+
+    gstate = Translate;
+    nextgState = Ready;
+
+
+    const int maxgpabits = SV39_LEVELS * SV39_LEVEL_BITS +
+                    SV39X4_WIDENED_BITS + PageShift;
+    Addr maxgpa = mask(maxgpabits);
+
+    if (debug_condition()) {
+        DPRINTF(PageTableWalker, "-------------------------");
+        DPRINTF(PageTableWalker,
+        " [GSTAGE BEGIN]: gpa: %#x\n", guest_paddr);
+    }
+    // If there is a bit beyond maxgpa, throw pf
+    if (guest_paddr & ~maxgpa) {
+        return pageFault();
+    }
+
+    // If there is another read packet,
+    // deallocate it, gstage creates a new packet
+    if (read) {
+        delete read;
+        read = nullptr;
+    }
+
+    Addr pte_addr = setupWalk(guest_paddr);
+    read = createReqPacket(pte_addr, MemCmd::ReadReq, sizeof(PTESv39));
+    glevel = SV39_LEVELS - 1;
+
+    // TODO THE TIMING PATH IS UNTESTED
+    if (timing)
+    {
```

Comment:
The style is not right in this block of code. E.g.,

```suggestion
    if (timing) {
```